File: fullstack-db-optimization-system-pro/backend/app/services/user_service.py
from typing import List, Optional
from sqlalchemy.ext.asyncio import AsyncSession
from app.db import crud
from app.db.models import User, UserRole
from app.schemas.user import UserCreate, UserUpdate
from app.core import security
from app.core.exceptions import NotFoundException, ConflictException, ForbiddenException
from loguru import logger

class UserService:
    async def get_user(self, db: AsyncSession, user_id: int) -> User:
        user = await crud.user.get(db, user_id)
        if not user:
            raise NotFoundException(f"User with ID {user_id} not found")
        return user

    async def get_user_by_email(self, db: AsyncSession, email: str) -> Optional[User]:
        return await crud.user.get_by_email(db, email)

    async def get_user_by_username(self, db: AsyncSession, username: str) -> Optional[User]:
        return await crud.user.get_by_username(db, username)

    async def get_all_users(self, db: AsyncSession, skip: int = 0, limit: int = 100) -> List[User]:
        return await crud.user.get_multi(db, skip=skip, limit=limit)

    async def create_user(self, db: AsyncSession, user_in: UserCreate, current_user: User) -> User:
        if current_user.role != UserRole.ADMIN:
            raise ForbiddenException("Only administrators can create new users directly.")

        existing_user_email = await crud.user.get_by_email(db, email=user_in.email)
        if existing_user_email:
            raise ConflictException(f"User with email '{user_in.email}' already exists.")

        existing_user_username = await crud.user.get_by_username(db, username=user_in.username)
        if existing_user_username:
            raise ConflictException(f"User with username '{user_in.username}' already exists.")

        hashed_password = security.get_password_hash(user_in.password)
        new_user = await crud.user.create_with_hashed_password(db, obj_in=user_in, hashed_password=hashed_password)
        logger.info(f"Admin '{current_user.username}' created user: {new_user.username} (ID: {new_user.id})")
        return new_user

    async def update_user(self, db: AsyncSession, user_id: int, user_in: UserUpdate, current_user: User) -> User:
        db_user = await self.get_user(db, user_id)

        # Only admins can update other users, or a user can update themselves
        if current_user.id != user_id and current_user.role != UserRole.ADMIN:
            raise ForbiddenException("You do not have permission to update this user.")

        # Admins can change roles and active status, regular users cannot change these for themselves
        if current_user.id == user_id and (user_in.is_admin is not None or user_in.role is not None):
            raise ForbiddenException("Users cannot change their own admin status or role.")
        
        # Prevent non-admins from changing other users' roles/active status
        if current_user.role != UserRole.ADMIN and current_user.id != user_id:
            if user_in.is_admin is not None or user_in.role is not None or user_in.is_active is not None:
                 raise ForbiddenException("Only administrators can update user roles or active status.")

        update_data = user_in.model_dump(exclude_unset=True)

        if "password" in update_data and update_data["password"]:
            update_data["hashed_password"] = security.get_password_hash(update_data.pop("password"))
        
        updated_user = await crud.user.update(db, db_user, update_data)
        logger.info(f"User '{current_user.username}' updated user ID: {updated_user.id}")
        return updated_user

    async def delete_user(self, db: AsyncSession, user_id: int, current_user: User) -> User:
        db_user = await self.get_user(db, user_id)

        if current_user.role != UserRole.ADMIN:
            raise ForbiddenException("Only administrators can delete users.")
        if current_user.id == user_id:
            raise ForbiddenException("You cannot delete your own user account.")

        deleted_user = await crud.user.remove(db, user_id)
        logger.info(f"Admin '{current_user.username}' deleted user: {deleted_user.username} (ID: {deleted_user.id})")
        return deleted_user

user_service = UserService()

File: enterprise-cms-app-complete/backend/src/services/userService.js
const { User } = require('../models');
const { ApiError } = require('../middleware/errorHandler');
const logger = require('../utils/logger');

/**
 * Get all users.
 * @param {object} queryOptions - Options for pagination, filtering, etc.
 * @returns {object} - Paginated list of users.
 */
const getAllUsers = async (queryOptions = {}) => {
  const { limit = 10, offset = 0, role, status } = queryOptions;

  const where = {};
  if (role) where.role = role;
  if (status) where.status = status;

  const users = await User.findAndCountAll({
    where,
    attributes: ['id', 'username', 'email', 'role', 'status', 'createdAt', 'updatedAt'],
    limit: parseInt(limit, 10),
    offset: parseInt(offset, 10),
    order: [['createdAt', 'DESC']],
  });

  logger.info(`Fetched ${users.rows.length} users.`);
  return users;
};

/**
 * Get a single user by ID.
 * @param {string} userId - The ID of the user.
 * @returns {object} - The user object.
 * @throws {ApiError} 404 if user not found.
 */
const getUserById = async (userId) => {
  const user = await User.findByPk(userId, {
    attributes: ['id', 'username', 'email', 'role', 'status', 'createdAt', 'updatedAt', 'lastLogin'],
  });

  if (!user) {
    throw new ApiError(404, 'User not found.');
  }
  logger.info(`Fetched user with ID: ${userId}`);
  return user;
};

/**
 * Update a user.
 * @param {string} userId - The ID of the user to update.
 * @param {object} updateData - Data to update (e.g., username, email, role, status).
 * @param {object} requestingUser - The user making the request.
 * @returns {object} - Updated user object.
 * @throws {ApiError} 403 if attempting to change admin role without being admin.
 * @throws {ApiError} 404 if user not found.
 * @throws {ApiError} 400 if email/username already exists.
 */
const updateUser = async (userId, updateData, requestingUser) => {
  const user = await User.findByPk(userId);

  if (!user) {
    throw new ApiError(404, 'User not found.');
  }

  // Prevent non-admins from changing roles or status of other users, or their own role to admin
  if (requestingUser.role !== 'admin') {
    if (userId !== requestingUser.id) {
      if (updateData.role || updateData.status) {
        throw new ApiError(403, 'Forbidden: Only admins can change other users\' roles or status.');
      }
    } else { // User is trying to update their own profile
      if (updateData.role && updateData.role !== requestingUser.role) {
        throw new ApiError(403, 'Forbidden: You cannot change your own role.');
      }
      if (updateData.status && updateData.status !== requestingUser.status) {
        throw new ApiError(403, 'Forbidden: You cannot change your own status.');
      }
    }
  }

  // Check for duplicate username/email if they are being updated
  if (updateData.email && updateData.email !== user.email) {
    const existingEmailUser = await User.findOne({ where: { email: updateData.email } });
    if (existingEmailUser) {
      throw new ApiError(400, 'User with this email already exists.');
    }
  }
  if (updateData.username && updateData.username !== user.username) {
    const existingUsernameUser = await User.findOne({ where: { username: updateData.username } });
    if (existingUsernameUser) {
      throw new ApiError(400, 'User with this username already exists.');
    }
  }

  // Sanitize update data for specific roles
  const allowedUpdates = {};
  if (requestingUser.role === 'admin') {
    Object.assign(allowedUpdates, updateData);
  } else {
    // Non-admins can only update certain fields for themselves or if specifically allowed
    if (updateData.username) allowedUpdates.username = updateData.username;
    if (updateData.email) allowedUpdates.email = updateData.email;
    // Password change is handled via a separate endpoint/flow usually, but here, it would be direct
    if (updateData.password) allowedUpdates.password = updateData.password;
  }

  await user.update(allowedUpdates);

  logger.info(`User with ID: ${userId} updated by ${requestingUser.email}`);
  return user;
};

/**
 * Delete a user.
 * @param {string} userId - The ID of the user to delete.
 * @param {object} requestingUser - The user making the request.
 * @throws {ApiError} 403 if a non-admin tries to delete an admin or themselves.
 * @throws {ApiError} 404 if user not found.
 */
const deleteUser = async (userId, requestingUser) => {
  const userToDelete = await User.findByPk(userId);

  if (!userToDelete) {
    throw new ApiError(404, 'User not found.');
  }

  // Admins cannot delete themselves or the last admin
  if (userToDelete.role === 'admin' && userId === requestingUser.id) {
    const adminCount = await User.count({ where: { role: 'admin' } });
    if (adminCount <= 1) {
      throw new ApiError(403, 'Forbidden: Cannot delete the last admin user, or self-delete as an admin.');
    }
  }

  // Non-admins cannot delete other users or admins
  if (requestingUser.role !== 'admin' && userId !== requestingUser.id) {
    throw new ApiError(403, 'Forbidden: Only admins can delete other users.');
  }
  if (requestingUser.role !== 'admin' && userToDelete.role === 'admin') {
    throw new ApiError(403, 'Forbidden: Cannot delete an admin user.');
  }

  await userToDelete.destroy();
  logger.info(`User with ID: ${userId} deleted by ${requestingUser.email}`);
  return { message: 'User deleted successfully.' };
};

module.exports = {
  getAllUsers,
  getUserById,
  updateUser,
  deleteUser,
};